openedx/core/djangoapps/course_apps/tasks.py

```diff
@@ -0,0 +1,57 @@
+"""
+This file contains celery tasks for course apps.
+"""
+
+from celery import shared_task
+from celery.utils.log import get_task_logger
+from edx_django_utils.monitoring import set_code_owner_attribute
+from opaque_keys.edx.keys import CourseKey
+from opaque_keys.edx.locator import LibraryLocator
+
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from openedx.core.djangoapps.course_apps.models import CourseAppStatus
+from openedx.core.djangoapps.course_apps.plugins import CourseAppsPluginManager
+
+log = get_task_logger(__name__)
+
+
+@shared_task(name='openedx.core.djangoapps.course_apps.tasks.cache_all_course_apps_status')
+@set_code_owner_attribute
+def cache_all_course_apps_status():
+    """
+    Create CourseAppStatus entries for all course apps, across all courses to speed up queries.
+    """
+    course_app_plugins = CourseAppsPluginManager.get_available_plugins()
+    for idx, course_key in enumerate(CourseOverview.objects.values_list('id', flat=True)):
+        if (idx + 1) % 100 == 0:
+            log.info("Cached %s records successfully", idx + 1)
+        if isinstance(course_key, LibraryLocator):
+            continue
+        for plugin in course_app_plugins.values():
+            # For current plugins this is just reading a django setting or returning a bool value
+            if plugin.is_available(course_key):
```

Comment:
Warning: We don't need to hold up this PR over this, but at some point, someone, somewhere is going to make their plugin make database or service calls to answer this question, and then the performance of this task is going to crater. Even though it might seem more wasteful, it's probably safer to have this task spin off new tasks for each individual course to do the work. This would also help keep intermittent errors (e.g. row lock timeout) from affecting other rows.

FWIW, enqueuing all courses into the task queue is done for course outline generation, and we haven't had trouble with it. It typically takes something like 20-30 mins to regenerate all course outlines, but that's a far more expensive operation than this would be doing.

Reply:
There was already a task for doing this for a single course at a time, so I've updated the bulk task to use that.